operators/src/engine/initialized_path.rs

```diff
@@ -0,0 +1,359 @@
+use async_trait::async_trait;
+
+use crate::{
+    error::Error,
+    util::{
+        input::{MultiRasterOrVectorOperator, RasterOrVectorOperator},
+        Result,
+    },
+};
+
+use super::{
+    ExecutionContext, InitializedRasterOperator, InitializedVectorOperator,
+    MultipleRasterOrSingleVectorSource, MultipleRasterSources, MultipleVectorSources,
+    SingleRasterOrVectorSource, SingleRasterSource, SingleVectorMultipleRasterSources,
+    SingleVectorSource,
+};
+
+#[derive(Debug, Clone, PartialEq, Eq, Hash, Default)]
+pub struct WorkflowOperatorPath {
+    id: Vec<u8>,
+}
+
+impl WorkflowOperatorPath {
+    pub fn new(id: Vec<u8>) -> Self {
+        Self { id }
+    }
+
+    pub fn inner(self) -> Vec<u8> {
+        self.id
+    }
+
+    #[must_use]
+    pub fn clone_and_extend(&self, suffix: &[u8]) -> Self {
+        let mut id = self.id.clone();
+        id.extend(suffix);
+        Self { id }
+    }
+
+    pub fn starts_with(&self, prefix: &[u8]) -> bool {
+        self.id.starts_with(prefix)
+    }
+}
+
+impl AsRef<[u8]> for WorkflowOperatorPath {
+    fn as_ref(&self) -> &[u8] {
+        &self.id
+    }
+}
+
+impl From<&[u8]> for WorkflowOperatorPath {
+    fn from(id: &[u8]) -> Self {
+        Self { id: id.to_vec() }
+    }
+}
+
+#[async_trait]
+pub trait InitializedSources<Initialized, E = Error> {
+    /// Initialize the operator(s) with a prefix
+    async fn initialize_sources(
+        self,
+        path: WorkflowOperatorPath,
+        context: &dyn ExecutionContext,
+    ) -> Result<Initialized, E>;
+}
+
+pub struct InitializedSingleRasterSource<R: InitializedRasterOperator> {
+    pub raster: R,
+    pub path: WorkflowOperatorPath,
```

Comment:
do you ever actually use this path? Otherwise you could always just directly return the raster/vector and save a line in the initialization of alle the operators for getting the operator out.

Reply:
it is not used yet. However, i think we should use it to let each operator tell its position when traversing the graph. So i vote to keep it until we have an idea how to go on.